cloud/scope/powervs_cluster.go

```diff
@@ -871,6 +873,329 @@ func (s *PowerVSClusterScope) createDHCPServer() (*string, error) {
+					return fmt.Errorf("error creating security group rule: %v", err)
+				}
+			}
+		}
+	}
+
+	return nil
+}
+
+// createVPCSecurityGroup creates a VPC security group.
+func (s *PowerVSClusterScope) createVPCSecurityGroup(spec infrav1beta2.SecurityGroup) (*string, error) {
+	s.Logger.Info("Creating security group", "name", *spec.Name)
+
+	options := &vpcv1.CreateSecurityGroupOptions{
+		VPC: &vpcv1.VPCIdentity{
+			ID: s.GetVPCID(),
+		},
+		Name: spec.Name,
+		ResourceGroup: &vpcv1.ResourceGroupIdentity{
+			ID: ptr.To(s.GetResourceGroupID()),
+		},
+	}
+
+	securityGroup, _, err := s.IBMVPCClient.CreateSecurityGroup(options)
+	if err != nil {
+		return nil, err
+	}
+	// To-Do: Add tags to VPC security group, need to implement the client for "github.com/IBM/platform-services-go-sdk/globaltaggingv1".
+
+	return securityGroup.ID, nil
+}
+
+// validateVPCSecurityGroupRuleRemote compares a specific security group rule's remote with the spec and existing security group rule's remote.
+func (s *PowerVSClusterScope) validateVPCSecurityGroupRuleRemote(originalSGRemote *vpcv1.SecurityGroupRuleRemote, expectedSGRemote infrav1beta2.SecurityGroupRuleRemote) (bool, error) {
+	var match bool
+
+	switch expectedSGRemote.RemoteType {
+	case infrav1beta2.SecurityGroupRuleRemoteTypeAny:
+		if originalSGRemote.CIDRBlock != nil && *originalSGRemote.CIDRBlock == "0.0.0.0/0" {
+			match = true
+		}
+	case infrav1beta2.SecurityGroupRuleRemoteTypeIP:
+		if originalSGRemote.Address != nil && *originalSGRemote.Address == *expectedSGRemote.IP {
+			match = true
+		}
+	case infrav1beta2.SecurityGroupRuleRemoteTypeCIDR:
+		cidrSubnet, err := s.IBMVPCClient.GetVPCSubnetByName(*expectedSGRemote.CIDRSubnetName)
+		if err != nil {
+			return false, fmt.Errorf("error retrieving vpc subnet by name '%s' for getting cidr block: %v", *expectedSGRemote.CIDRSubnetName, err)
+		}
+
+		if originalSGRemote.CIDRBlock != nil && cidrSubnet != nil && *originalSGRemote.CIDRBlock == *cidrSubnet.Ipv4CIDRBlock {
+			match = true
+		}
+	case infrav1beta2.SecurityGroupRuleRemoteTypeSG:
+		securityGroup, err := s.IBMVPCClient.GetSecurityGroupByName(*expectedSGRemote.SecurityGroupName)
+		if err != nil {
+			return false, fmt.Errorf("error retrieving id for resource group '%s': %v", *expectedSGRemote.SecurityGroupName, err)
+		}
+
+		if originalSGRemote.CRN != nil && securityGroup.Name != nil && *originalSGRemote.CRN == *securityGroup.CRN {
+			match = true
+		}
+	}
+
+	return match, nil
+}
+
+// validateSecurityGroupRule compares a specific security group's rule with the spec and existing security group's rule.
+func (s *PowerVSClusterScope) validateSecurityGroupRule(originalSecurityGroup vpcv1.SecurityGroup, direction infrav1beta2.SecurityGroupRuleDirection, protocol string, portMin, portMax int64, icmpCode, icmpType *string, remote infrav1beta2.SecurityGroupRuleRemote) (bool, error) {
+	var match bool
+	var err error
```

Comment:
How about declaring them in return parameters?

Reply:
Done